rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+Although not really a data store from the point of the view of our application, another place we use S3 is with our `big_messaging` module to get round the 256KB limit on the size of SQS messages. Here we write the payload of the message to S3 if it exceeds the maximum size, defaulting to including the payload within the message otherwise.
+
+## Problem
+
+There are a number of issues with the usage of VHS as intermediate storage:
+
+* Perhaps most importantly, the current solution of using DynamoDB and S3 is very costly, and it is likely other methods will work out much cheaper.
+
+* There is very little visibility into the intermediate stored data, and it is a pain to have to look up the DynamoDB ID followed by an S3 location to see a single item.
+
+* Clean up of S3 buckets is not always performed, with stray data often persisting after tearing down a particular pipeline (not necessarily a problem with VHS itself, but worth bearing in mind for future solutions).
+
+* There will be a performance penalty to making two sequential read or write requests to DynamoDB and S3 which other stores will not suffer from. Other databases will also make use of a cache to improve read performance, whilst the DynamoDB component of VHS will not be able to cache the data itself.
+
+* Although the VHS is used throughout our code and has proven very reliable, it still feels like a workaround to the limitations of DynamoDB which would not be present in other stores.
+
+There has already been some discussion about usage of a different store to VHS, and due to the work on the enricher requiring a new intermediate store we have an opportunity to try out an alternative. If we can improve on these issues above, in the future we could potentially backport this to other parts of the pipeline.
+
+## Requirements for Intermediate Store
+
+Taking into account the business logic of our services, and given the issues with the current VHS highlighted in [Problem](#problem), there are a number of requirements and desirable features for any alternative store:
+
+1. The ability to insert items into the store with some unique string ID and some specific integer version (set externally to the pipeline from our adapters), with a method to only store data when the version given is greater or equal to what is currently in the store.
+
+2. The ability to retrieve data from the store, where the latest data is always retrieved and the version is returned with the data.
+
+3. Happily stores data in our current range of around 8GB, and scaling beyond this for future proofing.
+
+4. Some form of disk persistence so that we do not need to perform a reindex from scratch in the case of the database going down.
+
+5. Cheaper overall cost than currently, including read / write requests, associated networks costs, and storage costs.
+
+6. Ideally improved performance.
+
+7. Ideally better visibility into the data.
+
```

Comment:
As we talk about consistency and ACID compliance, what do we see as the requirements in this regard for this store?

Reply:
Well I think we could work around not having any of these properties (it may be desirable in some cases as its all to do with trade offs of course, see e.g. [CAP theorem](https://en.wikipedia.org/wiki/CAP_theorem)). Elasticsearch is eventually consistent rather than atomically consistent for example, Redis not completely durable depending on persistence model used etc

Reply:
I'd agree that eventually consistant is fine for this. We have a 10 loose guideline as to when updates will be seen in the pipeline. Maybe worth us having a performance budget in place to work to in general.

This is pretty interesting https://research.google/pubs/pub39966/